### PR TITLE
evo fallback policy

### PR DESCRIPTION
**Pull request by VaclavK**
have a look **especially at response codes ** definition for transient etc.

### CODE REVIEW

**Comment by VaclavK** [comment]
note that we have story in the backlog to pinpoint specific behaviors of reverse proxy for load balancing etc. to contrast the reverse proxy and cluster (ELB) level (beside the basic use case of service instance being moved/recycled)

the MS doco is rather silent on this

**Comment by arturek** [comment]
from [docs](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reverseproxy)
When the target service is stateless, reverse proxy picks a random instance of the service partition to forward the request to

File: src/Eshopworld.Web/EvoFallbackPollyPolicyBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                                ctx.SetResponseCode(HttpStatusCode.RequestTimeout);
+                                break;
+                            case HttpOperationException opEx when opEx.Response != null:
+                                ctx.SetResponseCode(opEx.Response.StatusCode);
+                                break;
+                        }
+
+                        if (!ctx.IsTransient())
+                        {
+                            ctx.LevelUp(); //straight to next level
```
[the remark this answers is not in the document]
Comment (VaclavK) [reply src/Eshopworld.Web/EvoFallbackPollyPolicyBuilder.cs]: so as discussed on the call, response code is honored, transient gets repeated, so your 500 is the likely the most frequent scenario here and it will be treated as transient and repeated

agreed that **at this moment** the difference between RP and ELB is  not big but few comments from my  thinking around it
 - the **main motive** is to get successful response **if we can** so if there is an isolated issue (limited to small scope of instances) retry (even at ELB) has chance of succeeding or it will "fail faster" to FD where the set of instances is greater
 - ELB does better balancing and it also does probing so if we do return more meaningful probe results, ELB level is "better" to deal with this

I do not have good example of some meaningful repeatable response codes that I do not treat as transient (feel free to suggest them, anybody!) except may 409 (but that is quite questionable as it generally means different things to different services in terms of transient behaviour)

the transient definition as such is not "mine" - this is consistent with polly/autorest baked in policies

I am not opposed to any chance as long as we have concise understanding of target behaviour

**Comment by djfr** [comment]
Two general strategy comments:

- This policy should handle any response codes. We have a _pay to play_ principle within everything we own, so we should not try to build a generic retry policy for all `HttpClient` types. At the same time if we try to process response codes, we cross the boundary of what's purely related to the Cloud Network stack and what's behaviour towards the target system, that often will be a customer owned system, outside our control and often with custom retry behaviour.

- I don't think giving control over configuration of the policy is a good idea. Tweaks will be coming for general querying of telemetry and when we _finally_ have chaos tests in place where we can easily reproduce the edge cases this policy is designed for: ReverseProxy acting up without issuing a bad health report | LoadBalancer routing to a failed instance before 2 consecutive probes fail.
